File: c++/155_Min_Stack.cpp
#include <vector>

using namespace std;

class MinStack {
public:
  std::vector<std::pair<int,int>> myStack;
  MinStack() {
  }
  
  void push(int val) {
    if (myStack.empty()) {
      myStack.push_back({val, val});
    } else {
      int prevMin = myStack.back().second;
      myStack.push_back({val, min(prevMin, val)});
    }
  }
  
  void pop() {
    myStack.pop_back();
  }
  
  int top() {
    return myStack.back().first;
  }
  
  int getMin() {
    return myStack.back().second;
  }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: python/155_Min_Stack.py
class Min_Stack:

  def __init__(self):
    self.myStack = list()

  def push(self, val: int) -> None:
    if not self.myStack:
      self.myStack.append((val, val))
    else:
      prevMin = self.myStack[-1][-1]
      self.myStack.append((val, min(prevMin, val)))

  def pop(self) -> None:
    self.myStack.pop()

  def top(self) -> int:
    return self.myStack[-1][0]

  def getMin(self) -> int:
    return self.myStack[-1][-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()